packages/driver/src/multi-domain/communicator.ts

```diff
@@ -0,0 +1,186 @@
+ *
+ * The 'postMessage' method is used to send events to the spec bridge communicator, while
+ * the 'message' event is used to receive messages from the spec bridge communicator.
+ * All events communicating across origins are prefixed with 'cross:origin:' under the hood.
+ * See https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage for more details.
+ * @extends EventEmitter
+ */
+export class PrimaryOriginCommunicator extends EventEmitter {
+  private crossOriginDriverWindows: {[key: string]: Window} = {}
+  userInvocationStack?: string
+
+  /**
+   * The callback handler that receives messages from secondary origins.
+   * @param {MessageEvent.data} data - a reference to the MessageEvent.data sent through the postMessage event. See https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/data
+   * @param {MessageEvent.source} source - a reference to the MessageEvent.source sent through the postMessage event. See https://developer.mozilla.org/en-US/docs/Web/API/MessageEvent/source
+   * @returns {Void}
+   */
+  onMessage ({ data, source }) {
+    // check if message is cross origin and if so, feed the message into
+    // the cross origin bus with args and strip prefix
+    if (data?.event?.includes(CROSS_ORIGIN_PREFIX)) {
+      const messageName = data.event.replace(CROSS_ORIGIN_PREFIX, '')
+
+      // NOTE: need a special case here for 'bridge:ready'
+      // where we need to set the crossOriginDriverWindows to source to
+      // communicate back to the iframe
+      if (messageName === 'bridge:ready' && source) {
+        this.crossOriginDriverWindows[data.originPolicy] = source as Window
+      }
+
+      if (data?.data?.err) {
+        data.data.err = reifySerializedError(data.data.err, this.userInvocationStack as string)
+      }
+
+      this.emit(messageName, data.data, data.originPolicy)
+
+      return
+    }
+
+    debug('Unexpected postMessage:', data)
+  }
+
+  /**
+   * Events to be sent to the spec bridge communicator instance.
+   * @param {string} event - the name of the event to be sent.
+   * @param {any} data - any meta data to be sent with the event.
+   */
+  toAllSpecBridges (event: string, data?: any) {
+    debug('=> to all spec bridges', event, data)
+
+    const preprocessedData = preprocessForSerialization<any>(data)
+
+    // if user defined arguments are passed in, do NOT sanitize it.
+    if (data?.args) {
+      preprocessedData.args = data.args
+    }
+
+    // If there is no crossOriginDriverWindows, there is no need to send the message.
+    Object.values(this.crossOriginDriverWindows).forEach((win: Window) => {
+      win.postMessage({
+        event,
+        data: preprocessedData,
+      }, '*')
+    })
+  }
+
+  toSpecBridge (originPolicy: string, event: string, data?: any) {
+    debug('=> to spec bridge', originPolicy, event, data)
+
+    const preprocessedData = preprocessForSerialization<any>(data)
+
+    // if user defined arguments are passed in, do NOT sanitize it.
```

Comment:
```suggestion
    // if user defined arguments are passed in, do NOT sanitize them.
```

Reply:
Addressed in https://github.com/cypress-io/cypress/pull/21137/commits/6c58d3dae89db12414b05a5bc92851a640e35339